core/threaded/scheduler_GEDF_NP_CI.c

```diff
@@ -0,0 +1,547 @@
+ *
+ * @param number_of_workers Indicate how many workers this scheduler will be
+ *  managing.
+ * @param option Pointer to a `sched_params_t` struct containing additional
+ *  scheduler parameters.
+ */
+void lf_sched_init(
+    size_t number_of_workers, 
+    sched_params_t* params
+) {
+    DEBUG_PRINT("Scheduler: Initializing with %d workers", number_of_workers);
+    if(!init_sched_instance(&_lf_sched_instance, number_of_workers, params)) {
+        // Already initialized
+        return;
+    }
+
+    size_t queue_size = INITIAL_REACT_QUEUE_SIZE;
+    if (params != NULL) {
+        if (params->max_reactions_per_level != NULL) {
+            // Recalculate the queue size
+            queue_size = 0;
+            for (size_t i = 0; i <= _lf_sched_instance->max_reaction_level; i++) {
+                queue_size += params->max_reactions_per_level[i];
+            }
+        }
+    }
+    DEBUG_PRINT("Scheduler: Adopting a queue size of %u.", queue_size);
+
+    _lf_sched_instance->_lf_sched_array_of_mutexes =
+        (lf_mutex_t*)calloc(1, sizeof(lf_mutex_t));
+    lf_mutex_init(&_lf_sched_instance->_lf_sched_array_of_mutexes[0]);
+
+    // Reaction queue ordered first by deadline, then by level.
+    // The index of the reaction holds the deadline in the 48 most significant
+    // bits, the level in the 16 least significant bits.
+    _lf_sched_instance->_lf_sched_triggered_reactions = pqueue_init(
+        queue_size, in_reverse_order, get_reaction_index, get_reaction_position,
+        set_reaction_position, reaction_matches, print_reaction);
+    _lf_sched_instance->_lf_sched_transfer_reactions = pqueue_init(
+        queue_size, in_reverse_order, get_reaction_index, get_reaction_position,
+        set_reaction_position, reaction_matches, print_reaction);
+    // Create a queue on which to put reactions that are currently executing.
+    _lf_sched_instance->_lf_sched_executing_reactions = pqueue_init(
+        queue_size, in_reverse_order, get_reaction_index, get_reaction_position,
+        set_reaction_position, reaction_matches, print_reaction);
+
+    _lf_sched_threads_info = (_lf_sched_thread_info_t*)calloc(
+        _lf_sched_instance->_lf_sched_number_of_workers,
+        sizeof(_lf_sched_thread_info_t));
+
+    for (int i = 0; i < _lf_sched_instance->_lf_sched_number_of_workers; i++) {
+        _lf_sched_threads_info[i].output_reactions = pqueue_init(
+            queue_size, in_reverse_order, get_reaction_index,
+            get_reaction_position, set_reaction_position, reaction_matches,
+            print_reaction);
+    }
+}
+
+/**
+ * @brief Free the memory used by the scheduler.
+ *
+ * This must be called when the scheduler is no longer needed.
+ */
+void lf_sched_free() {
+    for (int i = 0; i < _lf_sched_instance->_lf_sched_number_of_workers; i++) {
+        pqueue_free(_lf_sched_threads_info[i].output_reactions);
+    }
+    // pqueue_free(_lf_sched_instance->_lf_sched_triggered_reactions); FIXME: This
```

Comment:
This, too, we should document in an issue.

Reply:
I agree that we should document this in an issue. I didn't get to thoroughly investigate this, but we should give this issue a closer look. In the previous design, calling `pqueue_free(reaction_q)` would cause the same type of "double free" errors if I remember correctly.

Reply:
OK, I created an issue for you: https://github.com/lf-lang/reactor-c/issues/41